File: JavaScript/AoC2025/days/day11.js
import fs, { rmSync } from "fs"
import path from "path"

export default class Day11 {
    constructor() {
        const filePath = path.resolve("../../Data/2025/day11.txt");
        const data = fs.readFileSync(filePath, "utf-8")
            .trim()
            .split('\r\n')
            .map(line => line.split(' '));
        this.cache = {};
        this.connections = {};
        for (const line of data) {
            const key = line[0].slice(0, -1);
            const value = line.slice(1);
            this.connections[key] = value;
        } 
    }

    printResults() {
        console.log("Day 11:");
        console.log(`Part 1: ${this.part1()}`);
        console.log(`Part 2: ${this.part2()}`);
    }

    part1() {
        return this.dfs('you', 'out');
    }

    part2() {
        return this.dfs('svr', 'fft') * this.dfs('fft', 'dac') * this.dfs('dac', 'out');
    }

    dfs(start, target) {
        if (start === target)
            return 1;
        if (start === 'out')
            return 0;
        if (`${start},${target}` in this.cache)
            return this.cache[`${start},${target}`];
        let total = 0;
        for (const c of this.connections[start]) {
            total += this.dfs(c, target);
        }
        this.cache[`${start},${target}`] = total;
        return total;
    }

}

File: Python/AoC2025/days/day11.py
class Day11:
    def __init__(self):
        with open("../../Data/2025/day11.txt") as f:
            self.data = [line.rstrip('\n').split(' ') for line in f]
        self.connections = {x[0][0:-1] : x[1:] for x in self.data}
        self.cache = {}

    def part_one(self):
        return self.follow_path('you', 'out')
    
    def part_two(self):
        a = self.follow_path('svr', 'fft')
        b = self.follow_path('fft', 'dac')
        c = self.follow_path('dac', 'out')
        return a * b * c

    def follow_path(self, input, end):
        if input == end:
            return 1
        elif input == 'out':
            return 0
        if (input, end) in self.cache:
            return self.cache[(input, end)]
        total = 0
        for c in self.connections[input]:
            total += self.follow_path(c, end)
        self.cache[(input, end)] = total
        return total

    def print_solution(self):
        print("Day 11:")
        print(f"Part 1: {self.part_one()}")
        print(f"Part 2: {self.part_two()}")